Guide/file-storage.markdown

```diff
@@ -271,21 +290,57 @@ renderForm :: Company -> Html
 renderForm company = formFor company [hsx|
     {(textField #name)}
 
-    <input
-        type="file"
-        name="logoUrl"
-        class="form-control-file"
-        accept="image/*"
-        data-preview="#logoUrlPreview"
-    />
+    {(fileField #logoUrl) {additionalAttributes = [("accept", "image/*"), ("data-preview", "#logoUrlPreview")]} }
 
     <img id="logoUrlPreview"/>
 
     {submitButton}
 |]
 ```
 
-On the "Edit.hs" file, it can be helpful to see the logo that has already been uploaded. To do this, change "<img id="logoUrlPreview"/>" to "<img id="logoUrlPreview" src={comapny.logoUrl}/>." This will allow the preview to show the existing logo, and also update to display any newly uploaded logos.
+On the "Edit.hs" file, it can be helpful to see the logo that has already been uploaded. To do this, change "<img id="logoUrlPreview"/>" to "<img id="logoUrlPreview" src={company.logoUrl}/>." This will allow the preview to show the existing logo, and also update to display any newly uploaded logos.
+
+### Required Uploads
+
+As notes above, the `logoUrl` must be a Maybe type, but there are cases where we want to ensure a file is uploaded as part of the record submission. We can add required to the form:
+
+```haskell
+renderForm :: Company -> Html
+renderForm company = formFor company [hsx|
+    {(textField #name)}
+
+    {(fileField #logoUrl) {required = True}}
+
+    {submitButton}
+|]
+```
+
+That's handy, however frontend validation is enough, we need to do server side validation as well. We can do that by
+slightly changing the order of our commands:
+
+```haskell
+action UpdateCompanyAction { companyId } = do
+    company <- fetch companyId
+    company
+        -- Now we first upload the image, and populate the `logoUrl` field.
+        |> uploadToStorage #logoUrl
+        >>= buildCompany
+        >>= ifValid \case
+        -- ...
+
+
+buildCompany company = company
+    |> fill @["name"]
+    |> validateField #name nonEmpty
+    |> validateField #logoUrl nonEmpty -- Validate that the logoUrl is not empty.
+    |> return -- We're inside an IO, so we need to `return`
```

Comment:
```suggestion
    |> pure -- We're inside an IO, so we need to use `pure`
```

IHP uses `pure` instead of `return` everywhere to avoid confusion with keywords in other languages

Reply:
Oh, I've started with `pure`, and changed to `return` :smile: